comit/src/expiries.rs

```diff
@@ -0,0 +1,631 @@
+//! Standardized COMIT Expiry Times
+//!
+//! The COMIT protocols rely on two expiry times in order for the swap to be
+//! atomic. One expiry for the alpha ledger and one for the beta ledger. The
+//! beta ledger expiry time must elapse before the alpha ledger expiry time.
+
+use crate::timestamp::{self, Timestamp};
+use time::{prelude::*, Duration};
+
+// TODO: Currently we ignore deploy for Erc20.
+
+/// Current time as a UNIX timestamp from the perspective of the implementer.
+///
+/// Intended for getting the current time from the underlying blockchain.
+pub trait CurrentTime {
+    fn current_time(&self) -> Timestamp;
+}
+
+/// Data that defines and manipulates expiry times.
+///
+/// Language note: We use the term 'expiries' (plural of expiry) to make it
+/// explicit that there are two expiry times however we refer it as singular
+/// since it is the struct, or pair of expiries, that we are referring to.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct Expiries<A, B> {
+    /// Configuration values for calculating transition periods.
+    config: Config,
+    /// Alpha blockchain connector for getting the current time.
+    alpha_connector: A,
+    /// Beta blockchain connector for getting the current time.
+    beta_connector: B,
+    /// The alpha ledger expiry time.
+    alpha_expiry: Timestamp,
+    /// The beta ledger expiry time.
+    beta_expiry: Timestamp,
+}
+
+impl<A, B> Expiries<A, B>
+where
+    A: CurrentTime,
+    B: CurrentTime,
+{
+    /// Create an expiries that will be useful for the duration of `ttl`.
```

Comment:
Please define `ttl` somewhere in the code. I think at the start of this file would be good.

Reply:
As discussed, the presence of a configurable `ttl` is dangerous.

When deciding the `time to act` for Alice and Bob. You mentioned that for Alice it's 1 hour and Bob 15 minutes.

You also mentioned that the absolute expiries are calculated and included in published order.

Now, if a `ttl` is configurable and set to 12 hours, then the time to act for Alice is actually `1 hour + 12 hours - age of order`. which could be 12 or 13 hours.

This invalidates the previous statement (Alice's time to act is 1 hour).

The issue stems from the fact that the absolute expiries are included in the published orders, which is a bad choice.

Instead, relative expiries should be included in the published orders, allowing the validity of an order to be orthogonal to the execution expiries.

The absolute expiries must be agreed and communicated when the `time to act` stopwatch starts for Alice. This is when `nectar` replies `ok` to an order being taken.
Then, Alice can confirm that the absolute expiries matches expectation and she then has the agreed `time to act` to fund.

I understand it means changing quite a bit of thing on the orderbook part but this is not part of this project :)